packages/nimble-components/src/table/specs/table-column-specs/table-column-formatted-text.md

```diff
@@ -176,60 +179,101 @@ Follows the [Column Type Philosophy](/packages/nimble-components/src/table/specs
 
-The API will be specified in a future update to this document. Below is **an example API** that leverages the native browser [`Intl.NumberFormat` API](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat/NumberFormat): this is intended to help visualize what the API _could_ look like, but isn't yet finalized.
+This column will display the `placeholder` when `typeof` the value is not `"number"` (i.e. if the value is `null`, `undefined`, not present, or has a different runtime data type). Note that IEE 754 numbers like Infinity, NaN, and -0 are type `"number"` so will be displayed how each formatter converts them. This will preserve values like `"∞"`, `"NaN"` and `"-0"`.
+
+##### Examples
 
 ```html
 <nimble-table>
-    <nimble-table-column-number
+    <nimble-table-column-number-text
+        field-name="numericTag"
+        placeholder="Not set"
+    >
+        Tag
+    </nimble-table-column-number-text>
+
+    <nimble-table-column-number-text
         field-name="count"
-        number-format-locales="en-US"
-        number-format-options-style="decimal"
-        number-format-options-use-grouping="false"
-        number-format-options-maximum-fraction-digits="0"
-        number-format-options-rounding-mode="trunc"
+        format="integer"
+        placeholder="Not found"
     >
         Count
-    </nimble-table-column-number>
+    </nimble-table-column-number-text>
 
-    <nimble-table-column-number
+    <nimble-table-column-number-text
         field-name="voltage"
-        number-format-locales="en-US"
-        number-format-options-style="decimal"
-        number-format-options-use-grouping="false"
+        format="custom"
+        locales="en-US"
+        custom-style="decimal"
+        custom-use-grouping="false"
         suffix=" V"
     >
         Voltage
-    </nimble-table-column-number>
+    </nimble-table-column-number-text>
 </nimble-table>
 ```
 
-##### Date Column
+#### Date Column
 
-Nimble could introduce `nimble-table-column-date-text` which maps date-time values to localized strings. The API will be specified in a future update to this document. It will need to consider cases like date formatting (both for locale and other reasons) and how to provide localized strings, possibly by exposing [the native browser `Intl.DateTimeFormat` API](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat).
+Nimble will introduce `nimble-table-column-date-text` which maps [JavaScript `Date`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date) values to localized strings.
 
----
-
-_These sections will be populated in a future update_
+Note that mutating a `Date` object does not cause the value to be rendered again.
 
-### API
+**Open question**: Should we represent dates using a numeric value (milliseconds since the Unix epoch) to avoid mutability concerns?
```

Comment:
I guess I'm not quite clear on the question. Do you mean should we expect people to supply values to represent as formatted Date strings, as numbers? If so, my personal opinion is no, as we've already tried to make it clear in the API that in order to update the display data in the table you have to call `setData(...)`.

Reply:
The concern I recall was that if records aren't just plain data then we have ambiguities about what happens if values are mutated. If I change the value of a date object in the data array, would the table observe that? What if I call `setData` again with the same object? Now should the table detect that the date has changed even though the reference is the same?

(For reference, this was the PR that removed it https://github.com/ni/nimble/pull/993)

Reply:
So, I believe one of the ideas of providing the `setData` method was to _remove_ the ambiguity of what happens when a user modifies a value in the array. There should be no assumptions by the end-user that the `Table` is using the _same instance_ of an object provided via the data given to `setData`. So, I believe (and I could be wrong) that the possible set of concerns are these:
- Does the table view update if a user mutates a `Date` instance?
    - I believe the answer to this will be "no" (which is what we want), as there is nothing observable happening here with respect to the table view
- If a user mutates a `Date` object, and then calls `setData` with the same data, will the table update with the new value?
    - I believe the answer here is "yes" (which is what we want), since the observable entity that will cause the cell to re-render (the `CellRecord` instance) will be new, as it is always re-generated when `setData` is called.

So, and it could take some validation, if I've captured the set of possible concerns above, and the answers to those concerns are accurate, would we want to reconsider supporting `Date`?

Reply:
> Does the table view update if a user mutates a Date instance?
> > I believe the answer to this will be "no"

I don't believe this is correct in the current implementation. We make a shallow copy of the records passed to setData so the Date references would still be held on to and mutable by the user unlike primitives which are copies:

https://github.com/ni/nimble/blob/14402b0d2e2a8a56454f79aa971324bf96dea5fe/packages/nimble-components/src/table/index.ts#L229

We would have to either add more logic to clone date objects or call `valueOf` on the data objects which results in a number representation (unix timestamp).

i don't think this benefits blazor. The serialization across the .NET context to the JS context would be much slower if we needed to support instantiating all the Date objects in the JS context. I'd like to ultimately get to something more akin to Apache Arrow data frames / flat buffers to represent the data across the context for performance / avoiding allocations.

> would we want to reconsider supporting Date?

It's not clear to me what are the arguments for reconsidering support of Date objects. Are there specific use cases you think are better served with Date objects?

Maybe supporting users that happen to have large arrays with Date objects in records already to make calling setData easier? In that case I think it should be isolated at setData and internally to the table we do not hold Date objects as the representation. It's not clear to me what the benefit is internally to the table.

I also don't think maximum ease of use of setData is something we are optimizing for. For example there are probably many ways to represent a url + a label, such as as markdown text. We aren't going to make the table / setData understand that or have a plugin to understand that. One needs to do pre-work to parse that into two named fields in a record and configure the column to understand it.

If there are convenience apis or functions to convert data for setData easier or to make setData "smarter" at autoconverting some data to the table formats we can consider it. I don't think we should plumb non-primitive column values through the table without a clear reason though.

Reply:
My main push is simply that the `Date` object is a fairly ubiquitous construct for representing time in JS, just as `DateTime` is in .NET. There is no doubt in my mind that not supporting `DateTime` directly in the Blazor `NimbleTable` will be seen as a major issue. I could even see it being seen as "table stakes".

That said, I believe you are correct in checking me on my assumption that the Table wouldn't display a mutated value of the `Date` instance.  Simply scrolling the table should result in the mutated `Date` object (not passed explicitly to `setData`) being displayed in the Table, which is not correct. So, yeah, I don't think we can support `Date` without doing burdensome work.

Reply:
@mollykreis, hmmm...that's something we should definitely investigate.

Reply:
Looked into this a bit. For both Blazor Server + Blazor Client/WASM, if the table data contains a C# `DateTime` object, the C# side (JsonSerializer) will serialize it as a string like `"2023-06-01T16:28:28.962-05:00"`.  
Converting that to a number should be possible on both the C# side (->`DateTimeOffset`, `.ToUnixTimeMilliseconds()`) and the JS side (it looks like the JS `Date()` parses that string format fine).  
Probably converting it on the C# side (before we send it to JS) makes the most sense.

Reply:
I updated the HLD to propose that we use numbers for the date column instead of Dates and added a section summarizing this thread.

This resulted in a slight change to the `placeholder` behavior: now `null` is treated the same as 0 because `new Date(null)` (somewhat surprisingly to me) produces a valid date object. We could special case this if someone feels strongly that `null` should display the `placeholder` value.